File: src/namagomi/minecraft/api/mods/NamagomiMod.ts
import {None, Option, Some} from "fp-ts/Option";

export interface NamagomiMod{
    side: 'SERVER' | 'CLIENT' | ''
    fileName: string
    downloadUrl: Some<string>  | None
    curseForge: Option<CurseForge>
}

export interface CurseForge{
    id: string
    gameVersion: string
    fileId: string
    hashes: Hash[]
}

export interface Hash{
    value: string
    algo: number
}

File: src/main/scala/com/github/namagomi/NamagomiMod.scala
package com.github.namagomi

type Side = "SERVER" | "CLIENT" | ""

sealed trait NamagomiMod

case class NamagomiModA
(
  name: Option[String],
  directUrl: String,
  side: Side
) extends NamagomiMod
case class NamagomiModB
(
  name: Option[String],
  modId: String,
  mcVersion: String,
  fileId: String,
  side: Side
) extends NamagomiMod